src/Sarif.Multitool/Rules/SARIF2004.OptimizeFileSize.cs

```diff
@@ -41,20 +41,94 @@ public class OptimizeFileSize : SarifValidationSkimmerBase
+        /// size of SARIF log files.
+        /// 
+        /// Prefer the result object properties 'ruleId' and 'ruleIndex' to the nested object-valued
+        /// property 'result.rule', unless the rule comes from a tool component other than the driver
+        /// (in which case only 'result.rule' can accurately point to the metadata for the rule).
+        /// The 'ruleId' and 'ruleIndex' properties are shorter and just as clear.
+        /// 
+        /// Do not specify the result object's 'analysisTarget' property unless it differs from the
+        /// result location. The canonical scenario for using 'result.analysisTarget' is a C/C++ language
+        /// analyzer that is instructed to analyze example.c, and detects a result in the included file
+        /// example.h. In this case, 'analysisTarget' is example.c, and the result location is in example.h.
         /// </summary>
         public override MultiformatMessageString FullDescription => new MultiformatMessageString { Text = RuleResources.SARIF2004_OptimizeFileSize_FullDescription_Text };
 
         protected override IEnumerable<string> MessageResourceNames => new string[] {
+            nameof(RuleResources.SARIF2004_OptimizeFileSize_Warning_AvoidDuplicativeAnalysisTarget_Text),
+            nameof(RuleResources.SARIF2004_OptimizeFileSize_Warning_AvoidDuplicativeResultRuleInformation_Text),
             nameof(RuleResources.SARIF2004_OptimizeFileSize_Warning_EliminateLocationOnlyArtifacts_Text),
-            nameof(RuleResources.SARIF2004_OptimizeFileSize_Warning_EliminateIdOnlyRules_Text)
+            nameof(RuleResources.SARIF2004_OptimizeFileSize_Warning_EliminateIdOnlyRules_Text),
+            nameof(RuleResources.SARIF2004_OptimizeFileSize_Warning_PreferRuleId_Text)
         };
 
         public override FailureLevel DefaultLevel => FailureLevel.Warning;
 
+        private string driverGuid;
+
         protected override void Analyze(Run run, string runPointer)
         {
             AnalyzeLocationOnlyArtifacts(run, runPointer);
             AnalyzeIdOnlyRules(run, runPointer);
+
+            this.driverGuid = run.Tool.Driver?.Guid;
+        }
+
+        protected override void Analyze(Result result, string resultPointer)
+        {
+            ReportUnnecessaryAnalysisTarget(result, resultPointer);
+            ReportRuleDuplication(result, resultPointer);
+        }
+
+        private void ReportRuleDuplication(Result result, string resultPointer)
+        {
+            if (result.Rule != null)
+            {
+                if (result.Rule.ToolComponent != null)
+                {
+                    if (result.Rule.ToolComponent.RefersToDriver(this.driverGuid))
+                    {
+                        // {0}: This result specifies both 'result.ruleId' and 'result.rule'.
+                        // Prefer 'result.ruleId' because it is shorter and just as clear.
+                        LogResult(
```

Comment:
>LogResult [](start = 24, length = 9)

This logic is backwards. Take a look at the notes I attached to the chat. The idea is:

If toolComponent is present, and if it does _not_ refer to the driver -- which means that it _has_ to be present! -- then `result.ruleId` and `result.ruleIndex` should not _also_ be present.

Here's what the notes said:
```
  if (result.Rule.ToolComponent != null && !result.Rule.ToolComponent.RefersToDriver(this.currentDriverGuid))
  {
    // We HAD to use result.Rule (because otherwise we couldn't have specified the extension component).
    if either result.ruleId or result.ruleIndex is present
        // Complain, because we should have put them in one place: under result.Rule along with the toolComponent.
        "This result provides the 'rule' property, but it also provides either or both of 'ruleId' and 'ruleIndex'. If 'rule' is present, then use 'rule.id' instead of 'ruleId', and use 'rule.index' instead of 'ruleIndex'."
  }
```